### ISSUE TITLE
Amok crashes when opening chrome's developer tools

### ISSUE THREAD

**Issue by jmventura** [caspervonb/amok#165]
hi there, every time I open the chrome's developer tools when running Amok it blows and I get this error:

`2015-12-08 22:54 node[33590] (FSEvents.framework) FSEventStreamFlushSync(): failed assertion '(SInt64)last_id > 0LL'`

Amok was launched by this command: `amok --browser chrome --hot index.js`

If I keep the devtools closed it works just fine.

My env is:
OS: OSX El Capitan version 10.11.1 (15B42)
Chrome: Version 47.0.2526.80 (64-bit)
Node: Version 4.2.1

Thanks for your nice work.

**Comment by caspervonb** [comment]
> 2015-12-08 22:54 node[33590] (FSEvents.framework) FSEventStreamFlushSync(): failed assertion '(SInt64)last_id > 0LL'

That's from chrome, no stack traces from node dumped? (should complain about a null socket unless its a new case)

**Comment by jmventura** [comment]
it always start with this stack:

```
λ ~/code/amok-test/ amok --browser chrome --hot index.js
2015-12-08 23:34:43.347 Google Chrome[37669:11341621] NSWindow warning: adding an unknown subview: <FullSizeContentView: 0x7f9e5373e9d0>. Break on NSLog to debug.
2015-12-08 23:34:43.348 Google Chrome[37669:11341621] Call stack:
(
	0   AppKit                              0x00007fff9c854807 -[NSThemeFrame addSubview:] + 107
	1   AppKit                              0x00007fff9c85454c -[NSView addSubview:positioned:relativeTo:] + 211
	2   Google Chrome Framework             0x0000000107affa43 ChromeMain + 51652483
	3   Google Chrome Framework             0x0000000107afe89c ChromeMain + 51647964
	4   Google Chrome Framework             0x0000000107b6b5d1 ChromeMain + 52093713
	5   Google Chrome Framework             0x0000000107abc77c ChromeMain + 51377340
	6   Google Chrome Framework             0x0000000107ac6d22 ChromeMain + 51419746
	7   Google Chrome Framework             0x0000000107bd6b4e ChromeMain + 52533390
	8   Google Chrome Framework             0x0000000107c0fc4a ChromeMain + 52767114
	9   Google Chrome Framework             0x0000000107c0f055 ChromeMain + 52764053
	10  Google Chrome Framework             0x0000000107c0ea82 ChromeMain + 52762562
	11  Google Chrome Framework             0x0000000107c0e1bd ChromeMain + 52760317
	12  Google Chrome Framework             0x0000000107c0dd0d ChromeMain + 52759117
	13  Google Chrome Framework             0x0000000107c0bd8f ChromeMain + 52751055
	14  Google Chrome Framework             0x0000000107c0bb13 ChromeMain + 52750419
	15  Google Chrome Framework             0x0000000107c0afb1 ChromeMain + 52747505
	16  Google Chrome Framework             0x0000000104a5d77f ChromeMain + 656575
	17  Google Chrome Framework             0x0000000104a5c69a ChromeMain + 652250
	18  Google Chrome Framework             0x0000000107e5bd4a ChromeMain + 55175818
	19  Google Chrome Framework             0x000000010810abf7 ChromeMain + 57989431
	20  Google Chrome Framework             0x0000000107e5a401 ChromeMain + 55169345
	21  Google Chrome Framework             0x0000000107e5e20d ChromeMain + 55185229
	22  Google Chrome Framework             0x0000000107e57f4f ChromeMain + 55159951
	23  Google Chrome Framework             0x0000000104f5045c ChromeMain + 5845404
	24  Google Chrome Framework             0x0000000104f4fa96 ChromeMain + 5842902
	25  Google Chrome Framework             0x00000001049bd302 ChromeMain + 66
	26  Google Chrome                       0x00000001049aef39 main + 9
	27  Google Chrome                       0x00000001049aef24 Google Chrome + 3876
)
```

it works just fine until I open the devtools, then boom:

```
2015-12-08 23:34 node[37668] (FSEvents.framework) FSEventStreamFlushSync(): failed assertion '(SInt64)last_id > 0LL'
```

After that the app continues to work on the browser, of course, but changes in the source doesn't get patched anymore and no other error arise.

**Comment by jmventura** [comment]
UPDATE: the issue only happens when installing Amok from npm (1.1.3).

Now I've installed it from master and works fine, but I had to change the event listener from `patch` to `scriptChange` in order to trigger the UI update. I've also noticed it opens a separate tab with the dev tools, is it by design?

**Comment by caspervonb** [comment]
So for amok to disconnect is here expected, **TL;DR:**

Navigate to `http://localhost:9222` and click your target there instead of clicking inspect on the page.

-------

The error is just a bit different than what i would expect. It should crash but it should crash with JavaScript error and a stack trace not C++ assertion failures. Anyhow I digress.

The reason amok disconnects is because chrome forces it to. Debugging sessions cannot be shared with multiple clients. Amok is a client and chrome is a client. Chrome of course takes priority and just kills any current connection when you open embedded devtools via "inspect element".

To workaround this, amok ships with a websocket multiplexer that lets both devtools and amok (and firefox devtools if you prefer those) connect via a single connection to the target browser.

Poorly documented, I know. There's this blog post https://medium.com/@.caspervonb/amok-and-external-developer-tools-9bcbddfd8f45

**Comment by jmventura** [comment]
Good to know then, I'll take your advice as it works for now, thanks a lot.

Just for curiosity, what's in your roadmap? I find Amok's approach much better then using eval...

**Comment by caspervonb** [comment]
In no particular order, the things my post-it notes say.

- Node.js support.
It's in https://github.com/caspervonb/amok/pull/164. However that branch is like a hodgepot of hackery and needs to be cherry picked / re-done.

- With Node.js support comes a rewrite of the executable script, tests need to be re-written to provide proper coverage.

- With Node.js support the test suite for Amok.js is obsolete.

- With Node.js support the documentation is obsolete, it's already somewhat out of sync.

- The supporting modules need some TLC as-well, e.g I need to push releases on more or less each one, test coverage is subpar (e.g am currently polishing http://node-browser_process.js.org)

The protocol layer `node-rdbg` has zero coverage, and has known minor issues (e.g crashing instead of throwing an error when socket is null is likely coming from `ws` in there) 

- Decide what to do about the "plugin" API.

This is a bigger one, right now a "plugin" is essentially just initialisation "middleware", a function that is run before the "Runner/Harness" attempts to make a connection to its target client. It's just the bare minimum the CLI needs to function.

Since the next release will be a major release, I want to make sure to break everything now and not now and the again the next release.

- Transforms!

This is a big one, A while back I wrote a plugin for browserify called amokify. I wouldn't expect it to work now but its a proof of concept that we can do non-destructive re-evaluation as an additional layer to what hotpatching offers (read this for more detail, demo, etc https://medium.com/@caspervonb/amokify-and-an-interactive-programming-workflow-11f863aca2d0#.qyjhoib84)

Now I targeted ES5, and v8 is essentially compatible with ES6 now. So had to scrap the implementation and need to R&D again.

- Be better than browsersync

Because we have a debugger connection, and can record input events, screencasts, etc. We can do way better than browsersync in terms of its mutlicasting. This has to be after everything else, but it's ahead of us.


- Mobile

Waay down the line, don't even have a test device anymore.

--------------

So to summarise, there's a bunch of tedious maintaince work to do. The two major features that are inbound would be Node.js support and re-evaluation transforms.

**Comment by jmventura** [comment]
Hi Casper, thanks for the insights, the roadmap seems very promising. I'm quite interested in Amok, I think it has lots of potential, what I like the most is the fact it is using Chrome's debugging protocol instead of userland eval for hot patching the sources.
